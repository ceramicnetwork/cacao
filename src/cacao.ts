import { SiweMessage } from 'siwe'

export type Header = {
  t: string
}

export type Payload = {
  domain: string
  iss: string
  chainId: number
  aud: string
  version: string
  nonce: string
  iat: number
  nbf?: number
  exp?: number
  statement?: string
  requestId?: string
  resources?: string[]
}

export type Signature = {
  s: string
}

export type Cacao = {
  h: Header
  p: Payload
  s: Signature
}

export interface CacaoBlock {
  cacao: Cacao
  // cid: CID
  bytes: Uint8Array
}

class CACAO {
  cacao: Cacao
  constructor(siwe: string | SiweMessage) {
    const siweMessage = typeof siwe === 'string' ? new SiweMessage(siwe) : siwe
    this.cacao = this.fromSiweMessage(siweMessage)
  }

  // TODO: Implement this API
  // toCacaoBlock(): CacaoBlock {}

  private fromSiweMessage(siweMessage: SiweMessage): Cacao {
    return {
      h: {
        t: 'eip4361-eip191',
      },
      p: {
        domain: siweMessage.domain,
        exp: Date.parse(siweMessage.expirationTime),
        iat: Date.parse(siweMessage.issuedAt),
        iss: siweMessage.address,
        chainId: Number(siweMessage.chainId),
        aud: siweMessage.uri,
        version: siweMessage.version,
        nonce: siweMessage.nonce,
        nbf: Date.parse(siweMessage.notBefore),
        statement: siweMessage.statement,
        requestId: siweMessage.requestId,
        resources: siweMessage.resources,
      },
      s: {
        s: siweMessage.signature,
      },
    }
  }
}

export default CACAO
