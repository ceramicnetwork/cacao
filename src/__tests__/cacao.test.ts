import { Wallet } from '@ethersproject/wallet'
import { Cacao, CacaoBlock } from '../cacao'
import { SiweMessage } from '../siwe'

describe('Cacao', () => {
  const wallet = Wallet.fromMnemonic(
    'despair voyage estate pizza main slice acquire mesh polar short desk lyrics'
  )
  const address = wallet.address

  test('Can create and verify Cacao Block', async () => {
    const msg = new SiweMessage({
      domain: 'service.org',
      address: address,
      statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
      uri: 'did:key:z6MkrBdNdwUPnXDVD1DCxedzVVBpaGi8aSmoXFAeKNgtAer8',
      version: '1',
      nonce: '32891757',
      issuedAt: '2021-09-30T16:25:24.000Z',
      chainId: '1',
      resources: [
        'ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu',
        'https://example.com/my-web2-claim.json',
      ],
    })

    const signature = await wallet.signMessage(msg.toMessage())
    msg.signature = signature

    const cacao = Cacao.fromSiweMessage(msg)
    const block = await CacaoBlock.fromCacao(cacao)
    expect(block).toMatchSnapshot()

    expect(() => Cacao.verify(cacao)).not.toThrow()
  })

  test('Converts between Cacao and SiweMessage', () => {
    const msg = new SiweMessage({
      domain: 'service.org',
      address: address,
      statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
      uri: 'https://service.org/login',
      version: '1',
      nonce: '32891757',
      issuedAt: '2021-09-30T16:25:24.000Z',
      chainId: '1',
      resources: [
        'ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu',
        'https://example.com/my-web2-claim.json',
      ],
    })

    const cacao = Cacao.fromSiweMessage(msg)
    const siwe = SiweMessage.fromCacao(cacao)
    expect(siwe).toEqual(msg)
  })

  test('ok after exp if within phase out period', async () => {
    const fixedDate = new Date('2021-10-14T07:18:41Z')
    const msg = new SiweMessage({
      domain: 'service.org',
      address: address,
      statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
      uri: 'https://service.org/login',
      version: '1',
      nonce: '32891757',
      issuedAt: fixedDate.toISOString(),
      expirationTime: new Date(fixedDate.valueOf() + 5 * 1000).toISOString(),
      chainId: '1',
      resources: [
        'ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu',
        'https://example.com/my-web2-claim.json',
      ],
    })

    const signature = await wallet.signMessage(msg.toMessage())
    msg.signature = signature

    const cacao = Cacao.fromSiweMessage(msg)
    const expiredTime = new Date(fixedDate.valueOf() + 10 * 1000)
    expect(() =>
      Cacao.verify(cacao, { atTime: expiredTime, revocationPhaseOutSecs: 20, clockSkewSecs: 0 })
    ).not.toThrow()
  })

  test('fail after exp if after phase out period', async () => {
    const fixedDate = new Date('2021-10-14T07:18:41Z')
    const msg = new SiweMessage({
      domain: 'service.org',
      address: address,
      statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
      uri: 'https://service.org/login',
      version: '1',
      nonce: '32891757',
      issuedAt: fixedDate.toISOString(),
      expirationTime: new Date(fixedDate.valueOf() + 5 * 1000).toISOString(),
      chainId: '1',
      resources: [
        'ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu',
        'https://example.com/my-web2-claim.json',
      ],
    })

    const signature = await wallet.signMessage(msg.toMessage())
    msg.signature = signature

    const cacao = Cacao.fromSiweMessage(msg)
    const expiredTime = new Date(fixedDate.valueOf() + 10 * 1000)
    expect(() =>
      Cacao.verify(cacao, { atTime: expiredTime, revocationPhaseOutSecs: 1, clockSkewSecs: 0 })
    ).toThrow(`CACAO has expired`)
  })

  test('ok before IAT if within default clockskew', async () => {
    const fixedDate = new Date('2021-10-14T07:18:41Z')
    const msg = new SiweMessage({
      domain: 'service.org',
      address: address,
      statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
      uri: 'https://service.org/login',
      version: '1',
      nonce: '32891757',
      issuedAt: fixedDate.toISOString(),
      chainId: '1',
      resources: [
        'ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu',
        'https://example.com/my-web2-claim.json',
      ],
    })

    const signature = await wallet.signMessage(msg.toMessage())
    msg.signature = signature

    const cacao = Cacao.fromSiweMessage(msg)
    const OneMinbeforeIAT = new Date(fixedDate.valueOf() - 60 * 1000)
    expect(() => Cacao.verify(cacao, { atTime: OneMinbeforeIAT })).not.toThrow()
  })
})
