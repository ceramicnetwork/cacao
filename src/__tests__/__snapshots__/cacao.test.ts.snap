// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cacao Can create and verify Cacao Block 1`] = `
Block {
  "asBlock": [Circular],
  "bytes": Uint8Array [
    163,
    97,
    104,
    161,
    97,
    116,
    103,
    101,
    105,
    112,
    52,
    51,
    54,
    49,
    97,
    112,
    168,
    99,
    97,
    117,
    100,
    120,
    56,
    100,
    105,
    100,
    58,
    107,
    101,
    121,
    58,
    122,
    54,
    77,
    107,
    114,
    66,
    100,
    78,
    100,
    119,
    85,
    80,
    110,
    88,
    68,
    86,
    68,
    49,
    68,
    67,
    120,
    101,
    100,
    122,
    86,
    86,
    66,
    112,
    97,
    71,
    105,
    56,
    97,
    83,
    109,
    111,
    88,
    70,
    65,
    101,
    75,
    78,
    103,
    116,
    65,
    101,
    114,
    56,
    99,
    105,
    97,
    116,
    120,
    24,
    50,
    48,
    50,
    49,
    45,
    48,
    57,
    45,
    51,
    48,
    84,
    49,
    54,
    58,
    50,
    53,
    58,
    50,
    52,
    46,
    48,
    48,
    48,
    90,
    99,
    105,
    115,
    115,
    120,
    59,
    100,
    105,
    100,
    58,
    112,
    107,
    104,
    58,
    101,
    105,
    112,
    49,
    53,
    53,
    58,
    49,
    58,
    48,
    120,
    66,
    100,
    57,
    68,
    57,
    99,
    55,
    68,
    67,
    51,
    56,
    57,
    55,
    49,
    53,
    97,
    56,
    57,
    102,
    67,
    56,
    49,
    52,
    57,
    69,
    52,
    97,
    53,
    66,
    101,
    57,
    49,
    51,
    51,
    54,
    66,
    50,
    55,
    57,
    54,
    101,
    110,
    111,
    110,
    99,
    101,
    104,
    51,
    50,
    56,
    57,
    49,
    55,
    53,
    55,
    102,
    100,
    111,
    109,
    97,
    105,
    110,
    107,
    115,
    101,
    114,
    118,
    105,
    99,
    101,
    46,
    111,
    114,
    103,
    103,
    118,
    101,
    114,
    115,
    105,
    111,
    110,
    97,
    49,
    105,
    114,
    101,
    115,
    111,
    117,
    114,
    99,
    101,
    115,
    130,
    120,
    53,
    105,
    112,
    102,
    115,
    58,
    47,
    47,
    81,
    109,
    101,
    55,
    115,
    115,
    51,
    65,
    82,
    86,
    103,
    120,
    118,
    54,
    114,
    88,
    113,
    86,
    80,
    105,
    105,
    107,
    77,
    74,
    56,
    117,
    50,
    78,
    76,
    103,
    109,
    103,
    115,
    122,
    103,
    49,
    51,
    112,
    89,
    114,
    68,
    75,
    69,
    111,
    105,
    117,
    120,
    38,
    104,
    116,
    116,
    112,
    115,
    58,
    47,
    47,
    101,
    120,
    97,
    109,
    112,
    108,
    101,
    46,
    99,
    111,
    109,
    47,
    109,
    121,
    45,
    119,
    101,
    98,
    50,
    45,
    99,
    108,
    97,
    105,
    109,
    46,
    106,
    115,
    111,
    110,
    105,
    115,
    116,
    97,
    116,
    101,
    109,
    101,
    110,
    116,
    120,
    65,
    73,
    32,
    97,
    99,
    99,
    101,
    112,
    116,
    32,
    116,
    104,
    101,
    32,
    83,
    101,
    114,
    118,
    105,
    99,
    101,
    79,
    114,
    103,
    32,
    84,
    101,
    114,
    109,
    115,
    32,
    111,
    102,
    32,
    83,
    101,
    114,
    118,
    105,
    99,
    101,
    58,
    32,
    104,
    116,
    116,
    112,
    115,
    58,
    47,
    47,
    115,
    101,
    114,
    118,
    105,
    99,
    101,
    46,
    111,
    114,
    103,
    47,
    116,
    111,
    115,
    97,
    115,
    162,
    97,
    115,
    120,
    132,
    48,
    120,
    49,
    48,
    57,
    51,
    49,
    51,
    101,
    55,
    53,
    50,
    53,
    100,
    101,
    97,
    53,
    53,
    101,
    99,
    57,
    97,
    51,
    99,
    99,
    98,
    98,
    54,
    51,
    101,
    97,
    56,
    100,
    54,
    56,
    52,
    48,
    54,
    51,
    54,
    54,
    50,
    53,
    48,
    99,
    102,
    48,
    56,
    56,
    48,
    100,
    54,
    55,
    48,
    51,
    50,
    98,
    52,
    53,
    55,
    97,
    98,
    51,
    51,
    99,
    57,
    50,
    54,
    99,
    54,
    55,
    102,
    102,
    51,
    102,
    99,
    99,
    54,
    54,
    97,
    99,
    51,
    49,
    98,
    97,
    97,
    54,
    56,
    54,
    56,
    97,
    56,
    48,
    97,
    49,
    50,
    102,
    98,
    101,
    54,
    98,
    55,
    54,
    51,
    56,
    97,
    56,
    57,
    102,
    52,
    102,
    54,
    100,
    53,
    49,
    97,
    48,
    50,
    50,
    57,
    53,
    57,
    48,
    99,
    102,
    54,
    54,
    55,
    54,
    102,
    49,
    99,
    97,
    116,
    102,
    101,
    105,
    112,
    49,
    57,
    49,
  ],
  "cid": Object {
    "code": 113,
    "hash": Uint8Array [
      18,
      32,
      238,
      16,
      188,
      79,
      31,
      209,
      8,
      250,
      175,
      253,
      198,
      75,
      57,
      66,
      210,
      174,
      81,
      145,
      164,
      204,
      19,
      42,
      241,
      238,
      248,
      110,
      106,
      86,
      34,
      153,
      240,
      11,
    ],
    "version": 1,
  },
  "value": Object {
    "h": Object {
      "t": "eip4361",
    },
    "p": Object {
      "aud": "did:key:z6MkrBdNdwUPnXDVD1DCxedzVVBpaGi8aSmoXFAeKNgtAer8",
      "domain": "service.org",
      "iat": "2021-09-30T16:25:24.000Z",
      "iss": "did:pkh:eip155:1:0xBd9D9c7DC389715a89fC8149E4a5Be91336B2796",
      "nonce": "32891757",
      "resources": Array [
        "ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu",
        "https://example.com/my-web2-claim.json",
      ],
      "statement": "I accept the ServiceOrg Terms of Service: https://service.org/tos",
      "version": "1",
    },
    "s": Object {
      "s": "0x109313e7525dea55ec9a3ccbb63ea8d68406366250cf0880d67032b457ab33c926c67ff3fcc66ac31baa6868a80a12fbe6b7638a89f4f6d51a0229590cf6676f1c",
      "t": "eip191",
    },
  },
}
`;
